### PR TITLE
Quest scripts and AI events/scripts

### PR DESCRIPTION
**Pull request by mindphluxnet**
While not entirely complete - many script actions are still missing implementation - this handles:

- quest start and end scripts (broadcast texts, spells being cast, emotes ...)
- random idle broadcast texts (like for example "Come get your fresh bread!" from a wandering baker in Stormwind)
- on-aggro broadcast texts (like "You no take candle!") 
- monsters casting spells (on themselves) on spawn, aggro or death
- **NEW!** generic scripts (for example, Undercity inhabitants training on dummies)

Since many of these actions only run when certain conditions are met I've implemented a condition checker for most of the possible conditions.

**Note**: this does **not** handle or include any encounter scripting like escort quests. These need to be scripted separately. This only handles database scripts.

I've playtested this for a while and it seems to be working but please tell me if something's throwing errors or is just plainly not working (although oftentimes it's just not yet implemented).

Since the script data is from retail some scripts plainly don't do what they are supposed to do because stuff changed. Lots of playtesting and subsequent script fixing will be needed.

Database additions are from [vmangos/core](https://github.com/vmangos/core)

### CODE REVIEW

File: game/world/managers/objects/gameobjects/GameObjectManager.py [hunk 1]
```diff
@@ -119,7 +120,7 @@ def initialize_from_gameobject_template(self, gobject_template):
                 self.gobject_template.type == GameObjectTypes.TYPE_CAMERA or \
                 self.gobject_template.type == GameObjectTypes.TYPE_TRAP or \
                 self.gobject_template.type == GameObjectTypes.TYPE_CHEST:
-            self.lock = gobject_template.data0
+            self.lock = gobject_template.data0
```
Comment (GrenderG) [inline game/world/managers/objects/gameobjects/GameObjectManager.py:123]: Extra tab.

File: game/world/managers/objects/item/ItemManager.py [hunk 2]
```diff
@@ -69,7 +70,7 @@ def __init__(self,
         self.spell_stats = []
         self.lock = 0  # Unlocked (0)
         self.display_id = 0
-        self.loot_manager = None  # Optional
+        self.loot_manager = None  # Optional
```
Comment (GrenderG) [inline game/world/managers/objects/item/ItemManager.py:73]: Extra tab.

File: game/world/managers/objects/script/AIEventHandler.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                if event.event_type == CreatureAIEventTypes.AI_EVENT_TYPE_CAST_SPELL_ON_SPAWN:
+                    chance_roll = random.randint(0, 100)
+                    if chance_roll <= event.event_chance:
+                        script = WorldDatabaseManager.creature_ai_script_get_by_id(event.action1_script)
+                        if script:
+                            self.creature.script_handler.enqueue_ai_script(self.creature, script)
+                    break
+    
+    def on_enter_combat(self):        
+        self.creature.script_handler.reset() # reset any scripts that were queued before combat (e.g. on spawn)
```
Comment (GrenderG) [inline game/world/managers/objects/script/AIEventHandler.py:22]: Please make sure all comments follow the same structure (start with uppercase and end with dot). :pray: There are more comments that aren't following it but I'm just going to comment on this one.

Comment (mindphluxnet) [reply game/world/managers/objects/script/AIEventHandler.py]: I hopefully fixed them all in the latest commit.

**Comment by GrenderG** [comment]
Please take a look at the conflicts. :pray:

File: game/world/managers/objects/script/ScriptHandler.py [hunk 4]
```diff
@@ -61,7 +61,7 @@ def __init__(self, object):
         self.CREATURE_FLEE_TEXT = WorldDatabaseManager.BroadcastTextHolder.broadcast_text_get_by_id(1150)
 
     def handle_script(self, script):
-            if script.command in SCRIPT_COMMANDS:
+            if script.command in SCRIPT_COMMANDS:
```
Comment (GrenderG) [inline game/world/managers/objects/script/ScriptHandler.py:64]: Extra spaces!

File: game/world/managers/objects/script/ScriptHandler.py [hunk 5]
```diff
@@ -0,0 +1,927 @@
...
+
+    def handle_script_command_flee(self, script):
+        if script.source and script.source.is_alive:
+            if not script.source.unit_flags & UnitFlags.UNIT_FLAG_FLEEING:
+                script.source.unit_flags |= UnitFlags.UNIT_FLAG_FLEEING
+                script.source.set_uint32(UnitFields.UNIT_FIELD_FLAGS, script.source.unit_flags)
+
+                # I feel so dirty doing this but it's not working otherwise.
+                flee_text = self.CREATURE_FLEE_TEXT.male_text
+                flee_text = flee_text.replace('%s ', '')
```
Comment (GrenderG) [inline game/world/managers/objects/script/ScriptHandler.py:633]: `%s` should be replaced by the name of the source unit.

Comment (mindphluxnet) [reply game/world/managers/objects/script/ScriptHandler.py:633]: Yes, but the client doesn't do it even though it should, and if the server does it the text looks like:

Defias Henchman: Defias Henchman attempts to run away in fear!

I went though it with Fluglow and he couldn't find another way either.

Comment (GrenderG) [reply game/world/managers/objects/script/ScriptHandler.py:633]: I see, probably in 0.5.3 it worked differently as the name is always included by default.

Comment (ratkosrb) [reply game/world/managers/objects/script/ScriptHandler.py]: Early vanilla clients have the same issue. I remove the %s upon loading of broadcast_text if server is compiled for older build.

Comment (GrenderG) [reply game/world/managers/objects/script/ScriptHandler.py]: Oh, good to know, I guess we can change it too at the time of loading.

Comment (mindphluxnet) [reply game/world/managers/objects/script/ScriptHandler.py]: I just pushed a commit doing exactly that.

Comment (GrenderG) [reply game/world/managers/objects/script/ScriptHandler.py]: > I just pushed a commit doing exactly that.

Great! Can you please add the explanatory comment there too? So we don't forget why we are replacing it on loading? :D

Comment (mindphluxnet) [reply game/world/managers/objects/script/ScriptHandler.py]: Done.

File: utils/constants/MiscCodes.py [hunk 6]
```diff
@@ -911,3 +911,33 @@ class ReputationSourceGain(IntEnum):
...
+    TELE_TO_GM_MODE = 1
+    TELE_TO_NOT_LEAVE_TRANSPORT = 2
+    TELE_TO_NOT_LEAVE_COMBAT = 0x04
+    TELE_TO_NOT_UNSUMMON_PET = 0x08
+    TELE_TO_SPELL = 0x10
+    TELE_TO_FORCE_MAP_CHANGE = 0x20
+
+class CreatureAIEventTypes(IntEnum):
+    AI_EVENT_TYPE_OUT_OF_COMBAT = 1
+    AI_EVENT_TYPE_ON_LOW_HP = 2
```
Comment (GrenderG) [inline utils/constants/MiscCodes.py:935]: I think this should be renamed to `AI_EVENT_TYPE_HP` or similar. Seems to be used for events related to HP (not only low) according to VMaNGOS source code.
```cpp
    EVENT_T_HP                      = 2,                    // HPMax%, HPMin%, RepeatMin, RepeatMax
```

Comment (mindphluxnet) [reply utils/constants/MiscCodes.py]: Yeah I renamed it

**Pull request merged**